objects/src/assets/mod.rs

```diff
@@ -0,0 +1,420 @@
+        asset[2] = Felt::new(temp | AccountId::NON_FUNGIBLE_FAUCET_TAG as u64);
+
+        Ok(Self(asset))
+    }
+
+    // HELPER FUNCTIONS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an error if:
+    /// - The first bit of the asset is not ZERO.
+    /// - The 32 least significant bits of the 3rd element are not set to 2^31.
+    fn validate(&self) -> Result<(), AssetError> {
+        let first_bit = self.0[0].as_int() >> 63;
+        if first_bit != 0 {
+            return Err(AssetError::non_fungible_asset_invalid_first_bit());
+        }
+
+        let tag = self.0[2].as_int() as u32;
+        if tag != AccountId::NON_FUNGIBLE_FAUCET_TAG {
+            return Err(AssetError::NonFungibleAssetInvalidTag(tag));
+        }
+
+        Ok(())
+    }
+}
+
+impl Deref for NonFungibleAsset {
+    type Target = Word;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<NonFungibleAsset> for Word {
+    fn from(asset: NonFungibleAsset) -> Self {
+        asset.0
+    }
+}
+
+impl From<NonFungibleAsset> for [u8; 32] {
+    fn from(asset: NonFungibleAsset) -> Self {
+        let mut result = [0_u8; 32];
+        result[..8].copy_from_slice(&asset.0[0].as_int().to_le_bytes());
+        result[8..16].copy_from_slice(&asset.0[1].as_int().to_le_bytes());
+        result[16..24].copy_from_slice(&asset.0[2].as_int().to_le_bytes());
+        result[24..].copy_from_slice(&asset.0[3].as_int().to_le_bytes());
+        result
+    }
+}
+
+impl From<NonFungibleAsset> for Asset {
+    fn from(asset: NonFungibleAsset) -> Self {
+        Asset::NonFungible(asset)
+    }
+}
+
+impl TryFrom<Word> for NonFungibleAsset {
+    type Error = AssetError;
+
+    fn try_from(value: Word) -> Result<Self, Self::Error> {
+        let asset = Self(value);
+        asset.validate()?;
+        Ok(asset)
+    }
+}
+
+impl TryFrom<[u8; 32]> for NonFungibleAsset {
+    type Error = AssetError;
+
+    fn try_from(value: [u8; 32]) -> Result<Self, Self::Error> {
+        let word = parse_word(value)?;
+        Self::try_from(word)
+    }
+}
+
+impl fmt::Display for NonFungibleAsset {
+    fn fmt(&self, _f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        todo!()
+    }
+}
+
+// NON-FUNGIBLE ASSET DETAILS
+// ================================================================================================
+
+/// Details about a non-fungible asset.
+///
+/// Unlike [NonFungibleAsset] struct, this struct contains full details of a non-fungible asset.
+pub struct NonFungibleAssetDetails {
+    faucet_id: AccountId,
+    asset_data: String,
+}
+
+impl NonFungibleAssetDetails {
+    /// Returns asset details instantiated from the specified faucet ID and asset data.
+    ///
+    /// # Errors
+    /// Returns an error if the provided faucet ID is not for a non-fungible asset faucet.
+    pub fn new(faucet_id: AccountId, asset_data: String) -> Result<Self, AssetError> {
```

Comment:
It might be more flexible here (at zero-cost abstraction) to allow the user to create a NFA detail from anything that implements `Into<String>` (json objects, etc)
```suggestion
    pub fn new<D>(faucet_id: AccountId, asset_data: D) -> Result<Self, AssetError> where D: Into<String> {
```

Reply:
See my thoughts here https://github.com/0xPolygonMiden/miden-base/pull/2#discussion_r1050257831